torch/nested/_internal/ops.py

```diff
@@ -255,19 +267,461 @@ def jagged_binary_pointwise(func, *args, **kwargs):
+            f"and value.is_nested: {value.is_nested} instead."
+        )
+    if query.dtype != key.dtype or query.dtype != value.dtype:
+        raise ValueError(
+            f"Expected query, key, and value to have the same dtype, "
+            f"but got query.dtype: {query.dtype}, key.dtype: {key.dtype}, "
+            f"and value.dtype: {value.dtype} instead."
+        )
+    if query.device != key.device or query.device != value.device:
+        raise ValueError(
+            f"Expected query, key, and value to have the same device type, "
+            f"but got query.device: {query.device}, key.device: {key.device}, "
+            f"and value.device: {value.device} instead."
+        )
+    if query.dim() < 2 or key.dim() < 2 or value.dim() < 2:
+        raise ValueError(
+            f"Expected query, key, and value to all be  at least 2 dimensional, but got query.dim: "
+            f"{query.dim()}, key.dim: {key.dim()} and value.dim: {value.dim()} instead."
+        )
+    if query._ragged_idx != 2 or key._ragged_idx != 2 or value._ragged_idx != 2:
+        raise ValueError(
+            f"Expected query, key, and value to all be be jagged at dimension 2, but got query._ragged_idx: "
+            f"{query._ragged_idx}, key._ragged_idx: {key._ragged_idx} and value._ragged_idx: {value._ragged_idx} instead."
+        )
+    if attn_mask is not None:
+        # TODO: Figure out whether masks are actually supported for this layout or not
+        if attn_mask.dtype != torch.bool and attn_mask.dtype != query.dtype:
+            raise ValueError(
+                f"Expected attn_mask dtype to be bool or to match query dtype, but got attn_mask.dtype: "
+                f"{attn_mask.dtype}, and query.dtype: {query.dtype} instead."
+            )
+
+
+def _can_use_math_sdpa_jagged(
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    attn_mask: Optional[torch.Tensor] = None,
+    dropout_p=0.0,
+    is_causal=False,
+    scale=None,
+):
+    if (
+        not query.is_contiguous()
+        or not key.is_contiguous()
+        or not value.is_contiguous()
+    ):
+        raise ValueError("If inputs are nested tensors they must be contiguous.")
+    if is_causal:
+        raise ValueError(
+            "Nested tensors for query / key are not supported when is_causal=True."
+        )
+
+
+def _select_sdp_backend(query, key, value, attn_mask, dropout, is_causal):
+    if (
+        not flash_sdp_enabled()
+        and not mem_efficient_sdp_enabled()
+        and not math_sdp_enabled()
+    ):
+        return SDPBackend.ERROR
+
+    ordering = (
+        SDPBackend.FLASH_ATTENTION,
+        SDPBackend.EFFICIENT_ATTENTION,
+        SDPBackend.MATH,
+    )
+
+    params = SDPAParams(query, key, value, attn_mask, dropout, is_causal)
+
+    for backend in ordering:
+        if backend == SDPBackend.FLASH_ATTENTION:
+            if can_use_flash_attention(params):
```

Comment:
does anything need to change in `can_use_flash_attention()` to work with `SDPAParams` that contain jagged layout NTs for `query`, `key`, `value`?

similar question for `can_use_efficient_attention()`?

Reply:
yes, both functions need to be refactored and split into different checks, as @drisspg mentioned a while ago. I just didn't have the time for that yesterday, and given how much code was already coming out, I figured I'd just push most of it first